ledger/recovering-indexer-integration-tests/src/test/suite/scala/com/digitalasset/platform/indexer/RecoveringIndexerIntegrationSpec.scala

```diff
@@ -348,4 +339,68 @@ object RecoveringIndexerIntegrationSpec {
 
     private class StateUpdatesFailedException extends RuntimeException("State updates failed.")
   }
+
+  class PartyOnlyQueueWriteService(
+      ledgerId: String,
+      participantId: Ref.ParticipantId,
+      queue: BoundedSourceQueue[(Offset, Update)],
+      source: Source[(Offset, Update), NotUsed],
+  ) extends WritePartyService
+      with ReadService {
+
+    private val offset = new AtomicLong(0)
```

Comment:
It's a bit confusing to see this atomic used together with a mutable.Buffer that's not synchronized. I guess it's just for the convenience of the getAndIncrement?

Reply:
But I guess it's fine since the other option is to use @volatile on it.

Reply:
out of curiosity - what makes it safe to use mutable stuff here?

Reply:
Our tests basically submit a few requests sequentially so there is no concurrency